### PR TITLE
[megatron] feat: add dynamic context parallel scheduling

### PR DESCRIPTION
**Pull request by xiaoyao0115**
### What does this PR do?

Adds Dynamic Context Parallel (DCP) scheduling for the Megatron backend in verl.

DCP keeps the existing static CP topology and per-rank sequence budget, but lets each packed micro-batch use a smaller local CP group when the sequence lengths do not need the full CP size. This reduces unnecessary CP communication and padding work for long-tail sequence-length distributions.

### Usage

Enable DCP on top of an existing Megatron CP configuration:

```bash
actor_rollout_ref.actor.megatron.dynamic_context_parallel=True
```

When DCP is enabled, verl keeps the existing `context_parallel_size` and `max_seqlen_per_dp_cp_rank` settings, and lets the scheduler choose the local CP size per packed micro-batch.

### Main changes

- Integrates Megatron-Core's dynamic CP scheduler into verl micro-batch preparation.
- Routes packed tensors and local token ownership metadata through Megatron forward/backward.
- Supports local-token loss computation and compact DCP output routing.
- Adds Dynamic CP documentation.

### Benchmark

Qwen3-30B-A3B SFT benchmark:

- Setup: TP1 / PP1 / EP8 / static CP4 topology, BF16, `max_seqlen_per_dp_cp_rank=4096`
- Dataset: long-tail sequence lengths from 1K to 16K, matched batches for Fixed CP4 and Dynamic CP
- Throughput, steps 3-25: Fixed CP4 `209.97K tok/s`, Dynamic CP `337.16K tok/s`, `+60.6%`
- Loss, steps 1-25: Fixed CP4 mean `0.02104872`, Dynamic CP mean `0.02105354`, relative mean diff `0.0229%`

Fixed CP4 vs Dynamic CP Performance

Fixed CP4 vs Dynamic CP Loss

### Checklist

- [x] Added/updated tests
- [x] Added documentation
- [x] Validated performance and numerical alignment on Qwen3-30B-A3B

### CODE REVIEW

File: verl/utils/dynamic_cp_scheduler.py
Comment (ISEEKYAN) [inline verl/utils/dynamic_cp_scheduler.py:1]: is it possible that we import the core algorithm from megatorn-core instead of rewrite on verl?

File: verl/workers/engine/megatron/transformer_impl.py [hunk 1]
```diff
@@ -621,20 +623,73 @@ def forward_backward_batch(self, data: TensorDict, loss_function: Callable, forw
...
 
-        micro_batches, indices = prepare_micro_batches(
-            data=data,
-            dp_group=self.get_data_parallel_group(),
-            num_batches_divided_by=num_batches_divided_by,
-            same_micro_num_in_dp=True,
-            min_num_micro_batch=None,
-        )
+        dcp_routing_info = None
+        if self.engine_config.dynamic_context_parallel:
```
Comment (ISEEKYAN) [inline verl/workers/engine/megatron/transformer_impl.py:855]: it might be better to write this DCP process inside the prepare_micro_batches to make the impl as clean as possible

File: verl/workers/engine/megatron/transformer_impl.py [hunk 2]
```diff
@@ -902,16 +985,35 @@ def forward_step(
...
             )
         else:
-            if not isinstance(temperature, torch.Tensor):
+            if temperature is None:
+                temperature = torch.ones(input_ids.shape[0], device=input_ids.device, dtype=torch.float32)
+            elif not isinstance(temperature, torch.Tensor):
                 temperature = torch.tensor([temperature] * input_ids.shape[0], device=input_ids.device)
 
             temperature = temperature.to(torch.float32)
+            if temperature.numel() == 0:
```
Comment (ISEEKYAN) [inline verl/workers/engine/megatron/transformer_impl.py]: why do we have so many temperature situations?

File: verl/workers/utils/losses.py
Comment (ISEEKYAN) [inline verl/workers/utils/losses.py:1]: it is not permitted to modify the loss function for megatron's specific feature, because it is a universal and backend agnostic function, we must make our DCP free of changing losses

**Review by ISEEKYAN — COMMENTED** [review]
hi @xiaoyao0115, thanks for your contribution. I have left some comments, and it would be better if you could provide some perf and precision curve to prove the effectiveness of DCP

File: verl/utils/dynamic_cp_scheduler.py
Comment (xiaoyao0115) [reply verl/utils/dynamic_cp_scheduler.py:1]: Yes, agreed. I have updated the implementation so that verl imports `DefaultDynamicCPScheduler` from Megatron-Core to dicide how to select cp_size and pack. The remaining code in verl is for data routing

File: verl/workers/engine/megatron/transformer_impl.py (hunk 1 above)
Comment (xiaoyao0115) [reply verl/workers/engine/megatron/transformer_impl.py:855]: okay

File: verl/workers/utils/losses.py
Comment (xiaoyao0115) [reply verl/workers/utils/losses.py:1]: got it

File: verl/workers/engine/megatron/transformer_impl.py (hunk 2 above)
Comment (xiaoyao0115) [reply verl/workers/engine/megatron/transformer_impl.py]: fixed

File: verl/utils/dynamic_cp_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,1470 @@
...
+
+                for i in range(group_start_gpu):
+                    new_micro_batches[i] = micro_batches[i]
+                    new_exec_times[i] = exec_times[i]
+                    new_sample_ids_per_gpu[i] = sample_ids_per_gpu[i]
+
+                for i in range(group_start_gpu, group_end_gpu + needed_count + 1):
+                    new_micro_batches[i] = micro_batches[group_end_gpu]
+                    new_exec_times[i] = get_total_workload_fn(micro_batches[group_end_gpu][0], next_power)
+                    new_sample_ids_per_gpu[i] = sample_ids_per_gpu[group_end_gpu]
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply verl/utils/dynamic_cp_scheduler.py]: Fixed. The verl-local fill_empty_gpus implementation has been removed, since the scheduling algorithm is now delegated to Megatron-Core.

File: verl/workers/engine/megatron/transformer_impl.py [hunk 4]
```diff
@@ -605,7 +603,11 @@ def forward_backward_batch(self, data: TensorDict, loss_function: Callable, forw
         tu.assign_non_tensor(data, sp_size=self.engine_config.context_parallel_size)
 
         # compute num_tokens in global batch for loss normalization
-        batch_num_tokens = data["loss_mask"].sum().to(get_device_id())
+        loss_mask_data = data.get("loss_mask", None)
+        if loss_mask_data is not None:
+            batch_num_tokens = loss_mask_data.sum().to(get_device_id())
+        else:
+            batch_num_tokens = torch.tensor(1.0, device=get_device_id())
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply verl/workers/engine/megatron/transformer_impl.py]: Fixed.

File: verl/workers/engine/megatron/transformer_impl.py [hunk 5]
```diff
@@ -90,6 +99,21 @@ def __init__(
         self.optimizer_config = optimizer_config
         self.checkpoint_config = checkpoint_config
         assert self.engine_config.use_mbridge, "use_mbridge must be True"
+        if self.engine_config.dynamic_context_parallel:
+            if not self.engine_config.use_remove_padding:
+                raise ValueError("dynamic_context_parallel requires use_remove_padding=True")
+            if self.engine_config.use_fused_kernels:
```
Comment (ISEEKYAN) [inline verl/workers/engine/megatron/transformer_impl.py]: fused_kernel (linear CE) is vital to long context RL, do we have a plan to make it compatible?
We may prioritize some key compatible items such as P0(plan to support later) P1( no planed). I would suggest order P0s in:
- fused_kernel
- router replay
- mtp

File: verl/workers/engine/megatron/transformer_impl.py [hunk 6]
```diff
@@ -672,11 +702,21 @@ def forward_backward_batch(self, data: TensorDict, loss_function: Callable, forw
         self._distillation_use_topk_active = tu.get_non_tensor_data(data, key="distillation_use_topk", default=False)
         tu.assign_non_tensor(data, sp_size=self.engine_config.context_parallel_size)
 
+        if self.engine_config.dynamic_context_parallel:
+            if self._distillation_use_topk_active or tu.get_non_tensor_data(
+                data, key="distillation_only", default=False
+            ):
```
Comment (ISEEKYAN) [inline verl/workers/engine/megatron/transformer_impl.py]: it is better to centralize all the compromised features to a single place

Comment (xiaoyao0115) [reply verl/workers/engine/megatron/transformer_impl.py]: good catch! I'll make the changes

File: verl/workers/engine/megatron/transformer_impl.py (hunk 5 above)
Comment (xiaoyao0115) [reply verl/workers/engine/megatron/transformer_impl.py]: The latest revision now supports all three items:

- Fused linear CE now receives the per-microbatch local_cp_size and DCP router padding mask throughout THD preprocessing and output reconstruction. Scalar and uniform temperatures are supported; non-uniform per-sample temperatures require disabling fused kernels.
- Router replay R2 and R3 now gather and scatter routes using the current microbatch's dynamic CP group. moe_router_fusion remains explicitly unsupported because it bypasses the replay hooks.
- MTP now uses the CP group carried by packed_seq_params for rolling and loss communication, with the corresponding per-token normalization.

**Review by ISEEKYAN — APPROVED** [review]

**Review by ISEEKYAN — APPROVED** [review]

**Review by ISEEKYAN — DISMISSED** [review]

**Review by ISEEKYAN — APPROVED** [review]
